math/mathcore/src/RanluxppEngineImpl.cxx

```diff
@@ -0,0 +1,209 @@
+};
+
+} // end anonymous namespace
+
+namespace ROOT {
+namespace Math {
+
+template <int w, int p>
+class RanluxppEngineImpl {
+
+private:
+   uint64_t fState[9];     ///< State of the generator
+   unsigned fPosition = 0; ///< Current position in bits
+
+   static constexpr const uint64_t *kA = RanluxppData<p>::kA;
+   static constexpr unsigned kMaxPos = 9 * 64;
+
+   /// Produce next block of random bits
+   void Advance()
+   {
+      mulmod(kA, fState);
+      fPosition = 0;
+   }
+
+public:
+   /// Return the next random bits, generate a new block if necessary
+   uint64_t NextRandomBits()
+   {
+      if (fPosition + w > kMaxPos) {
+         Advance();
+      }
+
+      unsigned idx = fPosition / 64;
+      unsigned offset = fPosition % 64;
+      unsigned numBits = 64 - offset;
+
+      uint64_t bits = fState[idx] >> offset;
+      if (numBits < w) {
+         bits |= fState[idx + 1] << numBits;
+      }
+      bits &= ((1UL << w) - 1);
+
+      fPosition += w;
+
+      return bits;
+   }
+
+   /// Initialize and seed the state of the generator
+   void SetSeed(uint64_t s)
+   {
+      fState[0] = 1;
+      for (int i = 1; i < 9; i++) {
+         fState[i] = 0;
+      }
+
+      uint64_t a_seed[9];
+      // Skip 2 ** 96 states.
+      powermod(kA, a_seed, 1UL << 48);
+      powermod(a_seed, a_seed, 1UL << 48);
+      // Skip another s states.
+      powermod(a_seed, a_seed, s);
+      mulmod(a_seed, fState);
+
+      fPosition = 0;
+   }
+
+   /// Skip an arbitrary amount of random numbers without generating them
+   void Skip(uint64_t n)
+   {
+      uint64_t left = (kMaxPos - fPosition) / w;
+      if (n < left) {
+         // Just skip the next few entries in the currently available bits.
+         fPosition += n * w;
+         return;
+      }
+
+      n -= left;
+      // Need to advance and possibly skip over blocks.
+      uint64_t nPerState = kMaxPos / w;
+      uint64_t skip = (n / nPerState);
+
+      uint64_t a_skip[9];
+      powermod(kA, a_skip, skip + 1);
+      mulmod(a_skip, fState);
+
+      // Potentially skip numbers in the freshly generated block.
+      uint64_t remaining = n - skip * nPerState;
+      fPosition = remaining * w;
+   }
+};
+
+template <int p>
+RanluxppEngine<p>::RanluxppEngine(uint64_t seed) : fImpl(new RanluxppEngineImpl<52, p>)
+{
+   fImpl->SetSeed(seed);
+}
+
+template <int p>
+RanluxppEngine<p>::~RanluxppEngine() = default;
+
+template <int p>
+double RanluxppEngine<p>::Rndm()
+{
+   return (*this)();
+}
+
+template <int p>
+double RanluxppEngine<p>::operator()()
+{
+   // Get 52 bits of randomness.
+   uint64_t bits = fImpl->NextRandomBits();
+
+   // Construct the double in [1, 2), using the random bits as mantissa.
+   uint64_t exp = 0x3ff0000000000000;
+   double random;
+   uint64_t *pRandom = (uint64_t *)&random;
+   *pRandom = exp | bits;
+
+   // Shift to the right interval of [0, 1).
+   return random - 1;
```

Comment:
Maybe you prefer this? Up to you!

```suggestion
   union {
     double drandom;
     uint64_t irandom;
   } castUnion;
   castUnion.irandom = exp | bits;
   
   // Shift to the right interval of [0, 1).
   return castUnion.drandom - 1;
```

Reply:
I was hoping that I could just cast somehow, but haven't found a way. I went with the `union` but made it anonymous.